src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs

```diff
@@ -0,0 +1,1117 @@
+                yearGap -= 1;
+            }
+
+            int numberOfInterval = yearGap / interval;
+
+            DateTime previousOccurrenceMonth = start.AddYears(numberOfInterval * interval).DateTime;
+
+            //
+            // Find the first occurence date matched the pattern
+            // Only one day of week in the month will be matched
+            previousOccurrence = DateTime.MaxValue;
+
+            foreach (DayOfWeek day in pattern.DaysOfWeek)
+            {
+                DateTime occurrenceDate = NthDayOfWeekInTheMonth(previousOccurrenceMonth, pattern.Index, day);
+
+                if (occurrenceDate + start.TimeOfDay < previousOccurrence)
+                {
+                    previousOccurrence = occurrenceDate + start.TimeOfDay;
+                }
+            }
+
+            numberOfOccurrences = numberOfInterval + 1;
+        }
+
+        /// <summary>
+        /// Check whether the duration is shorter than the minimum gap between recurrence of days of week.
+        /// </summary>
+        /// <param name="duration">The time span of the duration.</param>
+        /// <param name="interval">The recurrence interval.</param>
+        /// <param name="daysOfWeek">The days of the week when the recurrence will occur.</param>
+        /// <param name="firstDayOfWeek">The first day of the week.</param>
+        /// <returns>True if the duration is compliant with days of week, false otherwise.</returns>
+        private static bool IsDurationCompliantWithDaysOfWeek(TimeSpan duration, int interval, IEnumerable<DayOfWeek> daysOfWeek, DayOfWeek firstDayOfWeek)
+        {
+            Debug.Assert(interval > 0);
+
+            if (daysOfWeek.Count() == 1)
+            {
+                return true;
+            }
+
+            // Shift to the first day of the week
+            DateTime date = DateTime.Today.AddDays(
+                RemainingDaysOfTheWeek(DateTime.Today.DayOfWeek, firstDayOfWeek));
+
+            DateTime prev = DateTime.MinValue;
+
+            TimeSpan minGap = TimeSpan.FromDays(DayNumberOfWeek);
+
+            for (int i = 0; i < DayNumberOfWeek; i++)
+            {
+                if (daysOfWeek.Any(day =>
+                    day == date.DayOfWeek))
+                {
+                    if (prev == DateTime.MinValue)
+                    {
+                        //
+                        // Find a occurrence for the first time
+                        prev = date;
+                    }
+                    else
+                    {
+                        TimeSpan gap = date - prev;
+
+                        if (gap < minGap)
+                        {
+                            minGap = gap;
+                        }
+
+                        prev = date;
+                    }
+                }
+
+                date += TimeSpan.FromDays(1);
+            }
+
+            //
+            // It may across weeks. Check the next week if the interval is one week.
```

Comment:
> week if the interval is one week

Can the duration not be 2 weeks, or perhaps 15 days?

Reply:
No, it can't. The time window duration should at least be shorter than the interval duration.

``` C#
//
// Time window duration must be shorter than how frequently it occurs
if (timeWindowDuration > intervalDuration || 
    !IsDurationCompliantWithDaysOfWeek(timeWindowDuration, pattern.Interval, pattern.DaysOfWeek, pattern.FirstDayOfWeek))
{
    paramName = $"{nameof(settings.End)}";

    reason = TimeWindowDurationOutOfRange;

    return false;
}
```

https://github.com/microsoft/FeatureManagement-Dotnet/blob/zhiyuanliang/recurring-time-window/src/Microsoft.FeatureManagement/FeatureFilters/Recurrence/RecurrenceEvaluator.cs#L225